File: mt-pricing-frontend/lib/api/endpoints/channel-mirror.ts
"use client";

import env from "@/lib/env";
import { createSupabaseBrowserClient } from "@/lib/supabase/client";

export type DiffStatus = "match" | "drift" | "missing" | "queued";
export type BuyBoxState = "own" | "competitor" | "none";

export interface FieldDiff {
  field: string;
  mt: string | null;
  live: string | null;
  status: DiffStatus;
  lang?: string | null;
  is_mono?: boolean;
}

export interface DiffSummary {
  match: number;
  drift: number;
  missing: number;
  queued: number;
}

export interface ChannelListing {
  id: string;
  product_sku: string;
  channel_code: string;
  external_id: string;
  buybox_state: BuyBoxState;
  buybox_pct_7d: string | null;
  stock_qty: number | null;
  rating: string | null;
  reviews_count: number | null;
  last_sync_at: string | null;
  diff_summary: DiffSummary;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface DiffResponse {
  channel_code: string;
  sku: string;
  external_id: string;
  diffs: FieldDiff[];
  summary: DiffSummary;
  fetched_at: string | null;
}

export interface SyncLogEntry {
  id: string;
  channel_code: string;
  product_sku: string | null;
  event_type: "pull" | "push" | "diff";
  ok: boolean;
  summary: string | null;
  duration_ms: number | null;
  created_at: string;
}

export interface PublishResponse {
  channel_code: string;
  sku: string;
  external_id: string;
  ok: boolean;
  accepted_fields: string[];
  rejected_fields: string[];
  message: string | null;
}

async function authedFetch<T>(path: string, init: RequestInit = {}): Promise<T> {
  const supabase = createSupabaseBrowserClient();
  const {
    data: { session },
  } = await supabase.auth.getSession();
  const headers = new Headers(init.headers);
  if (!headers.has("Content-Type") && init.body) {
    headers.set("Content-Type", "application/json");
  }
  if (session?.access_token) {
    headers.set("Authorization", `Bearer ${session.access_token}`);
  }
  const res = await fetch(`${env.NEXT_PUBLIC_BACKEND_URL}${path}`, {
    ...init,
    headers,
    cache: "no-store",
  });
  if (!res.ok) {
    let detail: unknown;
    try {
      detail = await res.json();
    } catch {
      /* noop */
    }
    const msg =
      typeof detail === "object" && detail && "detail" in detail
        ? JSON.stringify((detail as { detail: unknown }).detail)
        : res.statusText;
    throw new Error(msg);
  }
  if (res.status === 204) return undefined as T;
  return (await res.json()) as T;
}

export const channelMirrorApi = {
  listings: (channel: string, params: { cursor?: string | null; limit?: number } = {}) => {
    const q = new URLSearchParams();
    if (params.cursor) q.set("cursor", params.cursor);
    if (params.limit !== undefined) q.set("limit", String(params.limit));
    const qs = q.toString();
    return authedFetch<{ items: ChannelListing[]; cursor: { next: string | null; prev: string | null }; page_size: number; total: number | null }>(
      `/api/v1/channels/${channel}/listings${qs ? `?${qs}` : ""}`,
    );
  },
  diff: (channel: string, sku: string): Promise<DiffResponse> =>
    authedFetch<DiffResponse>(`/api/v1/channels/${channel}/${sku}/diff`),
  sync: (channel: string, sku: string): Promise<DiffResponse> =>
    authedFetch<DiffResponse>(`/api/v1/channels/${channel}/${sku}/sync`, { method: "POST" }),
  publish: (channel: string, sku: string, fields?: string[]): Promise<PublishResponse> =>
    authedFetch<PublishResponse>(`/api/v1/channels/${channel}/${sku}/publish`, {
      method: "POST",
      body: JSON.stringify(fields ? { fields } : {}),
    }),
  syncLog: (channel: string, limit = 5): Promise<SyncLogEntry[]> =>
    authedFetch<SyncLogEntry[]>(`/api/v1/channels/${channel}/sync-log?limit=${limit}`),
};

File: mt-pricing-backend/app/schemas/channel_mirror.py
"""Schemas Pydantic para Channel Mirror endpoints.

Alineado con `mt-pricing-frontend/app/(app)/canales/amazon-uae/page.tsx`:
- ``MirrorRow`` (frontend) ↔ ``FieldDiffResponse`` (backend).
- ``SYNC_LOG`` (frontend) ↔ ``SyncLogEntry`` (backend).
- "Estado del listing" card ↔ ``ChannelListingResponse``.
"""

from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from typing import Annotated, Any, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, StringConstraints

CHANNEL_CODE_REGEX = r"^[a-z0-9_]{2,64}$"
ChannelCodeStr = Annotated[
    str,
    StringConstraints(min_length=2, max_length=64, pattern=CHANNEL_CODE_REGEX),
]


DiffStatusLiteral = Literal["match", "drift", "missing", "queued"]
BuyBoxStateLiteral = Literal["own", "competitor", "none"]


class FieldDiffResponse(BaseModel):
    """Una fila del comparador campo a campo."""

    model_config = ConfigDict(extra="forbid")

    field: str
    mt: Any
    live: Any
    status: DiffStatusLiteral
    lang: Literal["ar"] | None = None
    mono: bool = False
    notes: list[str] = Field(default_factory=list)


class DiffSummary(BaseModel):
    """Cuenta agregada por status (banner del frontend)."""

    model_config = ConfigDict(extra="forbid")

    match: int = 0
    drift: int = 0
    missing: int = 0
    queued: int = 0


class ChannelListingResponse(BaseModel):
    """Listing emparejado MT ↔ canal — fila del listado paginado."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    product_sku: str
    channel_code: str
    external_id: str
    buybox_state: BuyBoxStateLiteral
    buybox_pct_7d: Decimal | None = None
    stock_qty: int | None = None
    rating: Decimal | None = None
    reviews_count: int | None = None
    last_sync_at: datetime | None = None
    diff_summary: DiffSummary = Field(default_factory=DiffSummary)
    is_active: bool = True
    created_at: datetime
    updated_at: datetime


class DiffResponse(BaseModel):
    """Respuesta de ``GET /diff`` y ``POST /sync``."""

    model_config = ConfigDict(extra="forbid")

    channel_code: str
    sku: str
    external_id: str
    diffs: list[FieldDiffResponse]
    summary: DiffSummary
    fetched_at: datetime | None = None


class PublishRequest(BaseModel):
    """Body opcional para ``POST /publish``."""

    model_config = ConfigDict(extra="forbid")

    fields: list[str] | None = Field(
        default=None,
        description="Lista de fields a publicar. Si es None, todo el canonical.",
    )
    reason: str | None = Field(default=None, max_length=512)


class PublishResponseModel(BaseModel):
    """Resultado del intento de publicación."""

    model_config = ConfigDict(extra="forbid")

    ok: bool
    submission_id: str | None = None
    accepted_fields: list[str] = Field(default_factory=list)
    rejected_fields: list[str] = Field(default_factory=list)
    message: str | None = None


class SyncLogEntry(BaseModel):
    """Una entrada del Sync log card."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    channel_code: str
    product_sku: str | None = None
    event_type: Literal["pull", "push", "diff"]
    ok: bool
    summary: str | None = None
    duration_ms: int | None = None
    created_at: datetime


__all__ = [
    "BuyBoxStateLiteral",
    "ChannelCodeStr",
    "ChannelListingResponse",
    "DiffResponse",
    "DiffStatusLiteral",
    "DiffSummary",
    "FieldDiffResponse",
    "PublishRequest",
    "PublishResponseModel",
    "SyncLogEntry",
]
